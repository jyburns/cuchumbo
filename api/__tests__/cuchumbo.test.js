const sut = require('../lib/cuchumbo');
const seedrandom = require('seedrandom');

const participantsFixture = require('./__fixtures__/participants');

const _getSeededRng = () => {
  return seedrandom('superSecretSecret' + 'notSecretRequestId');
};

test('builds formatted particpant object', () => {
  const formattedParticipants = sut.buildFormattedParticipants(participantsFixture.unformattedParticipants);

  expect(formattedParticipants).toStrictEqual(participantsFixture.formattedParticipants);
});

test('sorts participants by exclusion size first', () => {
  const orderedParticipants = sut.buildOrderedParticipants(participantsFixture.sortByExclusions);

  expect(orderedParticipants).toMatchSnapshot();
});

test('sorts participants alphabetically by email second', () => {
  const orderedParticipants = sut.buildOrderedParticipants(participantsFixture.sortByAlpha);

  expect(orderedParticipants).toMatchSnapshot();
});

test('builds participant map', () => {
  const participantMap = sut.buildParticipantMap(participantsFixture.formattedParticipants);

  expect(participantMap).toMatchSnapshot();
});

test('generates all viable cuchumbos given a data set', () => {
  const formattedParticipants = sut.buildFormattedParticipants(participantsFixture.unformattedParticipants);
  const orderedParticipants = sut.buildFormattedParticipants(formattedParticipants);
  const remainingParticipantEmails = new Set(orderedParticipants.map(participant => participant.email));

  const potentialCuchumbos = sut.generatePotentialCuchumbos(orderedParticipants, remainingParticipantEmails);

  expect(potentialCuchumbos).toMatchSnapshot();
});

test('builds result object from participant map', () => {
  const participantMap = sut.buildParticipantMap(participantsFixture.formattedParticipants);
  const resultsObject = sut.buildCuchumboResults(participantsFixture.drawObject, participantMap);

  expect(resultsObject).toMatchSnapshot();
});

test('throw when there are over maximum participants', () => {
  const rng = _getSeededRng();
  const longList = [...Array(101).keys()];

  const generateForLongList = () => {
    sut.generateCuchumbo(rng, longList);
  };

  expect(generateForLongList).toThrowErrorMatchingSnapshot();
});

test('throw when there are no viable solutions', () => {
  const rng = _getSeededRng();
  const participants = participantsFixture.overConstrainedParticipants;

  const attemptToGenerate = () => {
    sut.generateCuchumbo(rng, participants);
  }

  expect(attemptToGenerate).toThrowErrorMatchingSnapshot();
});

test('generate cuchumbo is a pure function', () => {
  const resultsObject1 = sut.generateCuchumbo(_getSeededRng(), participantsFixture.unformattedParticipants);
  const resultsObject2 = sut.generateCuchumbo(_getSeededRng(), participantsFixture.unformattedParticipants);
  const resultsObject3 = sut.generateCuchumbo(_getSeededRng(), participantsFixture.unformattedParticipants);

  expect(resultsObject1).toMatchSnapshot();
  expect(resultsObject1).toStrictEqual(resultsObject2);
  expect(resultsObject2).toStrictEqual(resultsObject3);
});

test('generate cuchumbo matches snapshot battery', () => {
  const rng = _getSeededRng();

  for(let i = 0; i < 5; i++) {
    const resultsObject = sut.generateCuchumbo(rng, participantsFixture.unformattedParticipants);
    expect(resultsObject).toMatchSnapshot();
  }

});

test('generates large cuchumbo in reasonable amount of time', () => {
  const rng = _getSeededRng();
  const largeParticipantList = participantsFixture.buildLargeParticipantList(9);

  sut.generateCuchumbo(rng, largeParticipantList);
});
