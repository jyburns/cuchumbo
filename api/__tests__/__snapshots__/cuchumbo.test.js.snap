// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds participant map 1`] = `
Map {
  "a@example.com" => "a",
  "b@example.com" => "b",
  "c@example.com" => "c",
  "d@example.com" => "d",
}
`;

exports[`builds result object from participant map 1`] = `
Array [
  Object {
    "sendTo": "a@example.com",
    "sendToName": "a",
    "targetEmail": "b@example.com",
    "targetName": "b",
  },
  Object {
    "sendTo": "b@example.com",
    "sendToName": "b",
    "targetEmail": "a@example.com",
    "targetName": "a",
  },
  Object {
    "sendTo": "c@example.com",
    "sendToName": "c",
    "targetEmail": "d@example.com",
    "targetName": "d",
  },
  Object {
    "sendTo": "d@example.com",
    "sendToName": "d",
    "targetEmail": "c@example.com",
    "targetName": "c",
  },
]
`;

exports[`generate cuchumbo is a pure function 1`] = `
Array [
  Object {
    "sendTo": "a@example.com",
    "sendToName": "a",
    "targetEmail": "d@example.com",
    "targetName": "d",
  },
  Object {
    "sendTo": "b@example.com",
    "sendToName": "b",
    "targetEmail": "c@example.com",
    "targetName": "c",
  },
  Object {
    "sendTo": "c@example.com",
    "sendToName": "c",
    "targetEmail": "b@example.com",
    "targetName": "b",
  },
  Object {
    "sendTo": "d@example.com",
    "sendToName": "d",
    "targetEmail": "a@example.com",
    "targetName": "a",
  },
]
`;

exports[`generate cuchumbo matches snapshot battery 1`] = `
Array [
  Object {
    "sendTo": "a@example.com",
    "sendToName": "a",
    "targetEmail": "d@example.com",
    "targetName": "d",
  },
  Object {
    "sendTo": "b@example.com",
    "sendToName": "b",
    "targetEmail": "c@example.com",
    "targetName": "c",
  },
  Object {
    "sendTo": "c@example.com",
    "sendToName": "c",
    "targetEmail": "b@example.com",
    "targetName": "b",
  },
  Object {
    "sendTo": "d@example.com",
    "sendToName": "d",
    "targetEmail": "a@example.com",
    "targetName": "a",
  },
]
`;

exports[`generate cuchumbo matches snapshot battery 2`] = `
Array [
  Object {
    "sendTo": "a@example.com",
    "sendToName": "a",
    "targetEmail": "d@example.com",
    "targetName": "d",
  },
  Object {
    "sendTo": "b@example.com",
    "sendToName": "b",
    "targetEmail": "c@example.com",
    "targetName": "c",
  },
  Object {
    "sendTo": "c@example.com",
    "sendToName": "c",
    "targetEmail": "b@example.com",
    "targetName": "b",
  },
  Object {
    "sendTo": "d@example.com",
    "sendToName": "d",
    "targetEmail": "a@example.com",
    "targetName": "a",
  },
]
`;

exports[`generate cuchumbo matches snapshot battery 3`] = `
Array [
  Object {
    "sendTo": "a@example.com",
    "sendToName": "a",
    "targetEmail": "c@example.com",
    "targetName": "c",
  },
  Object {
    "sendTo": "b@example.com",
    "sendToName": "b",
    "targetEmail": "d@example.com",
    "targetName": "d",
  },
  Object {
    "sendTo": "c@example.com",
    "sendToName": "c",
    "targetEmail": "b@example.com",
    "targetName": "b",
  },
  Object {
    "sendTo": "d@example.com",
    "sendToName": "d",
    "targetEmail": "a@example.com",
    "targetName": "a",
  },
]
`;

exports[`generate cuchumbo matches snapshot battery 4`] = `
Array [
  Object {
    "sendTo": "a@example.com",
    "sendToName": "a",
    "targetEmail": "d@example.com",
    "targetName": "d",
  },
  Object {
    "sendTo": "b@example.com",
    "sendToName": "b",
    "targetEmail": "c@example.com",
    "targetName": "c",
  },
  Object {
    "sendTo": "c@example.com",
    "sendToName": "c",
    "targetEmail": "b@example.com",
    "targetName": "b",
  },
  Object {
    "sendTo": "d@example.com",
    "sendToName": "d",
    "targetEmail": "a@example.com",
    "targetName": "a",
  },
]
`;

exports[`generate cuchumbo matches snapshot battery 5`] = `
Array [
  Object {
    "sendTo": "a@example.com",
    "sendToName": "a",
    "targetEmail": "d@example.com",
    "targetName": "d",
  },
  Object {
    "sendTo": "b@example.com",
    "sendToName": "b",
    "targetEmail": "a@example.com",
    "targetName": "a",
  },
  Object {
    "sendTo": "c@example.com",
    "sendToName": "c",
    "targetEmail": "b@example.com",
    "targetName": "b",
  },
  Object {
    "sendTo": "d@example.com",
    "sendToName": "d",
    "targetEmail": "c@example.com",
    "targetName": "c",
  },
]
`;

exports[`generates all viable cuchumbos given a data set 1`] = `
Set {
  "[[\\"a@example.com\\",\\"d@example.com\\"],[\\"b@example.com\\",\\"c@example.com\\"],[\\"c@example.com\\",\\"b@example.com\\"],[\\"d@example.com\\",\\"a@example.com\\"]]",
  "[[\\"a@example.com\\",\\"c@example.com\\"],[\\"b@example.com\\",\\"d@example.com\\"],[\\"c@example.com\\",\\"b@example.com\\"],[\\"d@example.com\\",\\"a@example.com\\"]]",
  "[[\\"a@example.com\\",\\"d@example.com\\"],[\\"b@example.com\\",\\"a@example.com\\"],[\\"c@example.com\\",\\"b@example.com\\"],[\\"d@example.com\\",\\"c@example.com\\"]]",
}
`;

exports[`sorts participants alphabetically by email second 1`] = `
Array [
  Object {
    "email": "b@example.com",
    "exclusions": Set {},
    "name": "b",
  },
  Object {
    "email": "d@example.com",
    "exclusions": Set {},
    "name": "d",
  },
  Object {
    "email": "a@example.com",
    "exclusions": Set {
      "b@example.com",
    },
    "name": "a",
  },
  Object {
    "email": "c@example.com",
    "exclusions": Set {
      "a@example.com",
      "d@example.com",
    },
    "name": "c",
  },
]
`;

exports[`sorts participants by exclusion size first 1`] = `
Array [
  Object {
    "email": "b@example.com",
    "exclusions": Set {},
    "name": "b",
  },
  Object {
    "email": "d@example.com",
    "exclusions": Set {},
    "name": "d",
  },
  Object {
    "email": "a@example.com",
    "exclusions": Set {
      "b@example.com",
    },
    "name": "a",
  },
  Object {
    "email": "c@example.com",
    "exclusions": Set {
      "a@example.com",
      "d@example.com",
    },
    "name": "c",
  },
]
`;

exports[`throw when there are no viable solutions 1`] = `"no combinations exist with the given the constraints"`;

exports[`throw when there are over maximum participants 1`] = `"too many participants, max is 9"`;
